ghost/email-service/lib/email-event-storage.js

```diff
@@ -1,40 +1,75 @@
-const {EmailDeliveredEvent, EmailOpenedEvent, EmailBouncedEvent, EmailUnsubscribedEvent, SpamComplaintEvent} = require('@tryghost/email-events');
+const {EmailDeliveredEvent, EmailOpenedEvent, EmailBouncedEvent, EmailTemporaryBouncedEvent, EmailUnsubscribedEvent, SpamComplaintEvent} = require('@tryghost/email-events');
 const moment = require('moment-timezone');
+const logging = require('@tryghost/logging');
 
 class EmailEventStorage {
     #db;
     #membersRepository;
+    #models;
 
-    constructor({db, membersRepository}) {
+    constructor({db, models, membersRepository}) {
         this.#db = db;
+        this.#models = models;
         this.#membersRepository = membersRepository;
     }
 
     listen(domainEvents) {
         domainEvents.subscribe(EmailDeliveredEvent, async (event) => {
-            await this.handleDelivered(event);
+            try {
+                await this.handleDelivered(event);
+            } catch (err) {
+                logging.error(err);
+            }
         });
 
         domainEvents.subscribe(EmailOpenedEvent, async (event) => {
-            await this.handleOpened(event);
+            try {
+                await this.handleOpened(event);
+            } catch (err) {
+                logging.error(err);
+            }
         });
 
         domainEvents.subscribe(EmailBouncedEvent, async (event) => {
-            await this.handlePermanentFailed(event);
+            try {
+                await this.handlePermanentFailed(event);
+            } catch (e) {
+                logging.error(e);
+            }
+        });
+
+        domainEvents.subscribe(EmailTemporaryBouncedEvent, async (event) => {
+            try {
+                await this.handleTemporaryFailed(event);
+            } catch (e) {
+                logging.error(e);
+            }
         });
 
         domainEvents.subscribe(EmailUnsubscribedEvent, async (event) => {
-            await this.handleUnsubscribed(event);
+            try {
+                await this.handleUnsubscribed(event);
+            } catch (e) {
+                logging.error(e);
+            }
         });
 
         domainEvents.subscribe(SpamComplaintEvent, async (event) => {
-            await this.handleComplained(event);
+            try {
+                await this.handleComplained(event);
+            } catch (e) {
+                logging.error(e);
+            }
         });
     }
 
     async handleDelivered(event) {
+        // To properly handle events that are received out of order (this happens because of polling)
+        // we only can set an email recipient to delivered if they are not already marked as failed
+        // Why handle this her?  An email can be 'delivered' and later have a delayed bounce event. So we need to prevent that delivered_at is set again.
         await this.#db.knex('email_recipients')
             .where('id', '=', event.emailRecipientId)
+            .whereNull('failed_at')
```

Comment:
🤔 What happens if we have a temporary failure, and then we later have that email delivered by mailgun on retry ? I think in that case we'll miss the delievered event which'll not be right. We could maybe update the check for both null failed_at or severity is temporary which will mean only ignore delivered after permanent failures, wdyt ?

Reply:
For temporary failures we don't set failed_at, so that should still work, right? 🤔

Reply:
Ah right 👌